transports/webrtc/src/tokio/transport.rs

```diff
@@ -0,0 +1,611 @@
+                    });
+                }
+            }
+        }
+
+        Poll::Pending
+    }
+
+    /// Constructs a [`Multiaddr`] for the given IP address that represents our listen address.
+    fn listen_multiaddress(&self, ip: IpAddr) -> Multiaddr {
+        let socket_addr = SocketAddr::new(ip, self.listen_addr.port());
+
+        socketaddr_to_multiaddr(&socket_addr, Some(self.config.fingerprint))
+    }
+}
+
+impl Stream for ListenStream {
+    type Item = TransportEvent<<Transport as libp2p_core::Transport>::ListenerUpgrade, Error>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        loop {
+            if let Some(closed) = self.report_closed.as_mut() {
+                // Listener was closed.
+                // Report the transport event if there is one. On the next iteration, return
+                // `Poll::Ready(None)` to terminate the stream.
+                return Poll::Ready(closed.take());
+            }
+
+            if let Poll::Ready(event) = self.poll_if_watcher(cx) {
+                return Poll::Ready(Some(event));
+            }
+
+            // Poll UDP muxer for new addresses or incoming data for streams.
+            match ready!(self.udp_mux.poll(cx)) {
+                UDPMuxEvent::NewAddr(new_addr) => {
+                    let local_addr =
+                        socketaddr_to_multiaddr(&self.listen_addr, Some(self.config.fingerprint));
+                    let send_back_addr = socketaddr_to_multiaddr(&new_addr.addr, None);
+
+                    let upgrade = upgrade::inbound(
+                        new_addr.addr,
+                        self.config.inner.clone(),
+                        self.udp_mux.udp_mux_handle(),
+                        self.config.fingerprint,
+                        new_addr.ufrag,
+                        self.config.id_keys.clone(),
+                    )
+                    .boxed();
+
+                    return Poll::Ready(Some(TransportEvent::Incoming {
+                        upgrade,
+                        local_addr,
+                        send_back_addr,
+                        listener_id: self.listener_id,
+                    }));
+                }
+                UDPMuxEvent::Error(e) => {
+                    self.close(Err(Error::UDPMux(e)));
+                }
+            }
+        }
+    }
+}
+
+/// A config which holds peer's keys and a x509Cert used to authenticate WebRTC communications.
+#[derive(Clone)]
+struct Config {
+    inner: RTCConfiguration,
+    fingerprint: Fingerprint,
+    id_keys: identity::Keypair,
+}
+
+impl Config {
+    /// Returns a new [`Config`] with the given keys and certificate.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there's no fingerprint with the SHA-256 algorithm (see
+    /// [`RTCCertificate::get_fingerprints`]).
+    fn new(id_keys: identity::Keypair, certificate: RTCCertificate) -> Self {
```

Comment:
Can we try and make our own certificate type that can randomly generate itself given a `Rng`? I see several benefits:

- We don't leak the `webrtc` lib in our public API.
- A deterministic rng will allow you to derive it based on the ID key or anything else.
- Convenient API for the tests to generate a random one.
- Deterministic certificates are - I think - impossible with `rcgen` so the code has to be written anyway, might as well make it easy for our users.

Thoughts?